dss/events/chunkedtask/_awstest.py

```diff
@@ -54,3 +83,62 @@ def run_one_unit(self) -> typing.Optional[bool]:
         else:
             self.state[0] += 1
             return None  # more work to be done.
+
+
+class SupervisorTask(Task[dict, bool]):
+    SPAWNED_TASK_KEY = "key"
+    TIMEOUT_KEY = "timeout"
+    DEFAULT_TIMEOUT = 30.0
+
+    def __init__(self, state: dict, runtime: Runtime) -> None:
+        self.state = state
+        self.runtime = runtime
+        self.last_checked = None  # type: typing.Optional[float]
+
+    @property
+    def expected_max_one_unit_runtime_millis(self) -> int:
+        return sys.maxsize
+
+    def get_state(self) -> dict:
+        return self.state
+
+    def run_one_unit(self) -> typing.Optional[bool]:
+        if SupervisorTask.SPAWNED_TASK_KEY not in self.state:
+            # start the job
+            task_id = self.runtime.schedule_work(AWSFastTestTask, [0, 5], True)
+            timeout = time.time() + SupervisorTask.DEFAULT_TIMEOUT
+
+            self.state[SupervisorTask.SPAWNED_TASK_KEY] = task_id
+            self.state[SupervisorTask.TIMEOUT_KEY] = timeout
+
+            return None
+
+        if time.time() > self.state[SupervisorTask.TIMEOUT_KEY]:
+            return False
+
+        if self.check_success_marker():
+            return True
+        return None
+
+    def check_success_marker(self):
+        raise NotImplementedError()
+
+
+AWS_SUPERVISOR_TEST_CLIENT_NAME = "supervisortest"
+
+
+class AWSSupervisorTask(SupervisorTask):
+    def __init__(self, state: dict, runtime: AWSRuntime) -> None:
+        super().__init__(state, runtime)
+
+    def check_success_marker(self):
+        # don't pound the filter logs API to a pulp.
+        if (self.last_checked is not None and
+                time.time() < self.last_checked + 1):
+            time.sleep(1)
```

Comment:
One unfortunate concern to watch out for is the CloudWatch Logs API limits for the number of concurrent readers and their request rates. As the number of supervisor tasks increases (e.g. if you spawn a supervisor task for each object copy >64MB), using CloudWatch Logs for inter-lambda IPC may not scale. This is something you could stress-test for in an ad hoc test.

http://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/cloudwatch_limits_cwl.html

Reply:
This is only for the unit test.  The production code path does not rely on cloudwatch.